### ISSUE TITLE
VS2015 fatal: unable to load DLL `sqlceme35.dll`

### ISSUE THREAD

**Issue by CADbloke** [ErikEJ/SqlCeToolbox#61]
Clicked on the Tools Menu item to open the toolbox ...
![sqlcetoolboxcrash](https://cloud.githubusercontent.com/assets/147329/15279176/debf6d28-1b64-11e6-9d5c-8675ba1e9d85.png)

Freezes then crashes VS2015. Uninstalled & re-installed the toolbox with the same result.

Win7 x64 VS2015 with latest updates.

I ran `Devenv.exe /log ...` and kept the logs, let me know if you want me to send them to you. (note to self: `Dropbox \ Dev Stuff`)

````
Problem signature:
  Problem Event Name:	CLR20r3
  Problem Signature 01:	devenv.exe
  Problem Signature 02:	14.0.25123.0
  Problem Signature 03:	56f22f32
  Problem Signature 04:	System.Data.SqlServerCe
  Problem Signature 05:	3.5.5692.0
  Problem Signature 06:	486146b7
  Problem Signature 07:	11e
  Problem Signature 08:	0
  Problem Signature 09:	System.DllNotFoundException
  OS Version:	6.1.7601.2.1.0.256.48
  Locale ID:	3081
  Additional Information 1:	0a9e
  Additional Information 2:	0a9e372d3b4ad19135b953a78882e789
  Additional Information 3:	0a9e
  Additional Information 4:	0a9e372d3b4ad19135b953a78882e789

Read our privacy statement online:
  http://go.microsoft.com/fwlink/?linkid=104288&clcid=0x0409

If the online privacy statement is not available, please read our privacy statement offline:
  C:\Windows\system32\en-US\erofflps.txt
````
````
System.DllNotFoundException was unhandled
  HResult=-2146233052
  Message=Unable to load DLL 'sqlceme35.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
  Source=System.Data.SqlServerCe
  TypeName=""
  StackTrace:
       at System.Data.SqlServerCe.NativeMethods.SafeRelease(IntPtr& ppUnknown)
       at System.Data.SqlServerCe.SqlCeConnection.ReleaseNativeInterfaces()
       at System.Data.SqlServerCe.SqlCeConnection.Dispose(Boolean disposing)
       at System.Data.SqlServerCe.SqlCeConnection.Finalize()
  InnerException: 
````

**Comment by CADbloke** [comment]
I strongly suspect it has something to do with https://github.com/ErikEJ/SqlCeToolbox/wiki/Known-issues but instead of graying out the menu it grayed out Visual Studio. The offending computer has SQL CE 4 (4.0.8876.1) on it but not SQL CE3.5 (obviously). I will await advice before installing CE3.5 (in case you want me to troubleshoot)

**Comment by ErikEJ** [comment]
You have some parts of SQL CE 3.5 SP1 on your system (ie a broken 3,5 SP1 Install) - assume you are not using daily builds, but latest version from VS Gallery? I will attempt to repro with latest daily bits.

**Comment by ErikEJ** [comment]
I tried installing 3.5 SP1 (iso 3.5 SP2) on my machine, but was unable to repro with the latest daily bits - maybe you could give them a try: http://vsixgallery.com/extension/41521019-e4c7-480c-8ea8-fc4a2c6f50aa/

**Comment by CADbloke** [comment]
the download link in the page you linked == "The resource you are looking for has been removed, had its name changed, or is temporarily unavailable."
and yes, latest version from the gallery. I will try to track the problem for you before I fix the install of CE3.5

**Comment by ErikEJ** [comment]
Hmmm.. I have no issues with the link.

You should be able to build from source with just VS 2015!?

Great if you could investigate what SQL CE 3.5 x86 binaries (versions and locations) are present on your system .. in particular System.data.sqlserverce.dll and sqlceme35.dll

**Comment by CADbloke** [comment]
sure, where should I look for those?

**Comment by ErikEJ** [comment]
On your C: drive! 
In GAC under Windows\Microsoft.NET and C:\program files (x86)

**Comment by CADbloke** [comment]
my C: is quite crowded and I didn't know which files. No binaries for SQL CE 3.5 (apart from an installation of PerfectDisk - plenty for SQL CE 4.

Neither of the files you mentioned are installed, I only have SQL CE 4 on this PC

**Comment by CADbloke** [comment]
I found the problems with the broken VS2015 build...

 SqlCeToolbox.DDEX35.csproj at line 95 is unresolved

````xml
<Reference Include="Microsoft.VisualStudio.Data.Framework, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
 ````

same in SqlCeToolbox.DDEXe.csproj (same line number too)

SqlCeToolbox.csproj at Line 61, these files are missing from the `..\lib\` folder - see https://github.com/ErikEJ/SqlCeToolbox/tree/master/src/GUI/lib

````xml
<Reference Include="Microsoft.ReportViewer.Common, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <SpecificVersion>False</SpecificVersion>
      <HintPath>..\lib\Microsoft.ReportViewer.Common.dll</HintPath>
      <Private>False</Private>
    </Reference>
    <Reference Include="Microsoft.ReportViewer.WinForms, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <SpecificVersion>False</SpecificVersion>
      <HintPath>..\lib\Microsoft.ReportViewer.WinForms.dll</HintPath>
      <Private>False</Private>
    </Reference>
````

**Comment by ErikEJ** [comment]
Re files: System.data.sqlserverce.dll and sqlceme35.dll is what you should be looking for...

I will look into the ability to build without VS 2015 asap...

**Comment by CADbloke** [comment]
ok, so I need both SQL CE 3.5 and SQL CE 4 installed? I don't have SQLCE3.5 installed  I am 100% ok with installing it but the docs at https://github.com/ErikEJ/SqlCeToolbox/wiki  say (my emphasis)

> The Toolbox requires **one** of these SQL Server Compact runtime MSIs to be installed...

also (just in case I confused you)

> I will look into the ability to build **with**~~out~~ VS 2015 asap...

**Comment by ErikEJ** [comment]
I meant "without VS 2010" - sorry.

In order to build you need both SQLCE 3.5 and 4.0, yes.

**Comment by ErikEJ** [comment]
I will update the build instructions on the readme.md to clarify build requirements

**Comment by CADbloke** [comment]
what about to install from the Visual Studio Gallery as an end-user (my original problem)? That's the wiki page I linked, it says you only need one version of CE, not both.

**Comment by ErikEJ** [comment]
As an end user, you only need runtimes for the database engine versions that you need to work with.

**Comment by ErikEJ** [comment]
Hey, I just checked in fixes for the reference issues and missing files for build with VS 2015 only

**Comment by CADbloke** [comment]
confirmed, now builds ok in Visual Studio. 

Still can't download the newest vsix from http://vsixgallery.com/extension/41521019-e4c7-480c-8ea8-fc4a2c6f50aa/ - download link is http://vsixgallery.com/extensions/41521019-e4c7-480c-8ea8-fc4a2c6f50aa/SQL%20Server%20Compact%2FSQLite%20Toolbox%20v4.5.45.vsix

**Comment by ErikEJ** [comment]
Sorry, I see what ypou mean now about vsix gallery - I assumed you were unable to even open the site.
You can use the feed lin: http://vsixgallery.com/extensions/41521019-e4c7-480c-8ea8-fc4a2c6f50aa/extension.vsix

**Comment by ErikEJ** [comment]
Thanks for the update on the ability to build in VS 2015, btw

**Comment by CADbloke** [comment]
no problem. I'll try the vsix in a minute - for infinite value of "minute" - I pressed F5 in VS2015 and now it is downloading symbols for ... well, everything...ever.

**Comment by CADbloke** [comment]
I gave up on F5 - the Internet in Australia has a ping time of about 5 years.

The new vsix still crashes VS2015, same as before.

I'll try F5 again - this could take a while

**Comment by ErikEJ** [comment]
I think you can disable it http://stackoverflow.com/questions/3817223/disable-vs-downloading-public-symbols

**Comment by CADbloke** [comment]
yeah, that's somewhat quicker, thanks. I don't see the SQLCE toolbox in the experimental VS instance so I can't debug it. :(

**Comment by ErikEJ** [comment]
Tools menu!

**Comment by CADbloke** [comment]
yeah but no - not there. Nothing in the settings either.

**Comment by ErikEJ** [comment]
And the affected code is in ExplorerControl.xaml.cs

**Comment by ErikEJ** [comment]
Windows, other?

**Comment by ErikEJ** [comment]
Set a breakpoint in the init meyhod in the package.cs

**Comment by CADbloke** [comment]
I set a breakpoint in `SqlCeToolboxPackage.cs` and it didn't hit it. VS said it won't hit it because there are no symbols loaded. I don't think I will live long enough to see all the symbols load. 

Just confirming, "SqlCeToolbox" should be the startup project?

**Comment by CADbloke** [comment]
and no, wasn't in View - Other Windows either.

**Comment by ErikEJ** [comment]
Are you in release mode, perhaps?

**Comment by CADbloke** [comment]
was in Debug, AnyCPU. Switched to Debug x86 and it worked. It crashed, same as before. No debug symbols.

SQLCE3.5 seems to be a hard dependency?

**Comment by ErikEJ** [comment]
Nope, it is not. No breakpoints hit, in ExplorerControl.xaml.cs, loaded event

**Comment by ErikEJ** [comment]
Can I see the logs, that you mentioned in your initial post?

**Comment by CADbloke** [comment]
sure. I just scrolled through them, they are just the activity logs
oh, Github doesn't like 7zip zip files (as in *.zip). 
https://dl.dropboxusercontent.com/u/5115091/SQLCEtoolboxVSlogs.7z
let me know if you don't have 7zip but the f7z is 91kB, the .zip is 317kB

Let me know if there are any other tests / logs I can help with. I will leave this PC "broken" for now - I have others I can use.

**Comment by ErikEJ** [comment]
🏁  Hey, thanks a million for helping with this - finally nailed it down, thanks to your detailed info and a walk with the dog 😄  - try build 4.5.52 from here: http://vsixgallery.com/extensions/41521019-e4c7-480c-8ea8-fc4a2c6f50aa/extension.vsix

**Comment by CADbloke** [comment]
Tested build 4.5.52 - still has the same problem, sorry. 

Does `System.Data.Common.DbProviderFactories.GetFactory("System.Data.SqlServerCe.3.5");` catch `System.DllNotFoundException`?

No worries about trying to track this down, thanks a million for making this. 

I fix a lot of bugs on long walks.

**Comment by ErikEJ** [comment]
Need to do a better test when I have the time...

**Comment by CADbloke** [comment]
"time", ah yes, I remember that luxury. Vaguely. No problems, let me know what else I can do to help.

**Comment by ErikEJ** [comment]
I was finally able to remove the newer 3.5 version from GAC, and was able to repro your issue. The conclusion: The Toolbox requires SQL Compact 3.5 SP2 installed, as 3.5 SP1 is not supported and has compatibility issues on newer OSes. I was not able to "catch" the fatal crash in any way.

**Comment by CADbloke** [comment]
Glad you found the root cause. I will install SQLCE3.5. Don't forget to tweak the wiki so the users know that SQLCE3.5 is  a prerequisite to running the Toolbox, not just optional if you happen to want to connect to to SQLCE. I only use SQLite, hence my lack of CE.3.5.

https://github.com/ErikEJ/SqlCeToolbox/wiki

> The Toolbox requires one of these SQL Server Compact runtime MSIs to be installed, if you want to connect to a SQL Server Compact database file:

**Comment by ErikEJ** [comment]
Have thought of a couple of ideas for a fix, reopening

**Comment by CADbloke** [comment]
ok, still haven't "fixed" the busted P so I'll leave it alone until I hear from you.

**Comment by ErikEJ** [comment]
The fix is up now in the latest daily (4.5.60): http://vsixgallery.com/extensions/41521019-e4c7-480c-8ea8-fc4a2c6f50aa/extension.vsix - would be very grateful if you could give at try! 

If you use my tools, I would be very grateful for a rating or review here:
https://visualstudiogallery.msdn.microsoft.com/0e313dfd-be80-4afb-b5e9-6e74d369f7a1/view/Reviews

**Comment by CADbloke** [comment]
WIN! :+1:  Launches and does SQLite all ok. I noticed the SQLCE 3.5 option was grayed out.

Fixed. Thanks.
